src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs

```diff
@@ -318,7 +318,27 @@ public JoinableTask RunAsync(Func<Task> asyncMethod)
     /// </remarks>
     public JoinableTask RunAsync(Func<Task> asyncMethod, JoinableTaskCreationOptions creationOptions)
     {
-        return this.RunAsync(asyncMethod, synchronouslyBlocking: false, creationOptions: creationOptions);
+        return this.RunAsync(asyncMethod, synchronouslyBlocking: false, parentToken: null, creationOptions: creationOptions);
+    }
+
+    /// <summary>
+    /// Invokes an async delegate on the caller's thread, and yields back to the caller when the async method yields.
+    /// The async delegate is invoked in such a way as to mitigate deadlocks in the event that the async method
+    /// requires the main thread while the main thread is blocked waiting for the async method's completion.
+    /// </summary>
+    /// <param name="asyncMethod">The method that, when executed, will begin the async operation.</param>
+    /// <param name="parentToken">An optional token that identifies one or more <see cref="JoinableTask"/> instances, typically in other processes, that serve as 'parents' to this one.</param>
+    /// <param name="creationOptions">The <see cref="JoinableTaskCreationOptions"/> used to customize the task's behavior.</param>
+    /// <returns>An object that tracks the completion of the async operation, and allows for later synchronous blocking of the main thread for completion if necessary.</returns>
+    /// <remarks>
+    /// <para>Exceptions thrown by the delegate are captured by the returned <see cref="JoinableTask" />.</para>
+    /// <para>When the delegate resumes from a yielding await, the default behavior is to resume in its original context
+    /// as an ordinary async method execution would. For example, if the caller was on the main thread, execution
+    /// resumes after an await on the main thread; but if it started on a threadpool thread it resumes on a threadpool thread.</para>
+    /// </remarks>
+    public JoinableTask RunAsync(Func<Task> asyncMethod, string? parentToken, JoinableTaskCreationOptions creationOptions)
```

Comment:
this is fine. Because it is actually used inside the RPC usage case, when there is no parentToken, the newly created JoinableTask will never have a chance to be a part of a dependency chain, so it would be completely a waste, and will inject a token to the call chain which has no UI thread context in the first place, so I wonder whether it is possible to prevent this in the first place.  Of course, this can be done in the RPC side, instead of something on JTF side, so feel free to ignore.

Reply:
Creating a JoinableTask without a `parentToken` is not a waste, because it may be that the incoming RPC call may itself make an outbound call, in which case we'd want a JoinableTask with its own token to use as a parent for the outbound call.

Reply:
yes, it can make further calls, but if it is not in a JTF context in the first place, it can never be joined/waited by a JTF context in the future (because the JTF task created here will be just an isolated one), so it will ensure the entire outbound chain has no UI thread dependencies (unless a middle link actively create a JTF and manages it, which will still work)  This will ensure we will not add any new overhead for call chains which have no UI thread links currently.

Reply:
the primary reason is that JTF instance is quite big, and the creation requires the global sync lock, which is a known contention point, so preventing doing that can prevent some of this overhead.

Reply:
I guess you're right. I can't think of any valid reason we'd need to wrap an incoming call with a JoinableTask if no parent token exists. I've taken this into account in my StreamJsonRpc PR.